### ISSUE TITLE
[ROCm] [Build Error] [caffe2] `cuda.h/cuda_runtime.h` file not found

### ISSUE THREAD

**Issue by acxz** [pytorch/pytorch#42430]
## 🐛 Bug

When compiling pytorch with ROCm, pytorch fails when trying to compiling caffe2 related code. Even with `BUILD_CAFFE2_OPS=0`

```
Building wheel torch-1.6.0a0+b31f58d
-- Building version 1.6.0a0+b31f58d
cmake --build . --target install --config Release -- -j 4
[1/231] Building HIPCC object caffe2/CMakeFiles/torch_hip.dir/utils/math/torch_hip_generated_broadcast.cu.o
FAILED: caffe2/CMakeFiles/torch_hip.dir/utils/math/torch_hip_generated_broadcast.cu.o
cd /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math && /usr/bin/cmake -E make_directory /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math/. && /usr/bin/cmake -D verbose:BOOL=OFF -D build_configuration:STRING=RELEASE -D generated_file:STRING=/home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math/./torch_hip_generated_broadcast.cu.o -P /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math/torch_hip_generated_broadcast.cu.o.cmake
In file included from /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/caffe2/utils/math/broadcast.cu:3:
In file included from /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/caffe2/core/context_gpu.h:8:
/home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/caffe2/core/common_gpu.h:5:10: fatal error: 'cuda.h' file not found
#include <cuda.h>
         ^~~~~~~~
1 error generated when compiling for host.
CMake Error at torch_hip_generated_broadcast.cu.o.cmake:138 (message):
  Error generating
  /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math/./torch_hip_generated_broadcast.cu.o


[2/231] Building HIPCC object caffe2/CMakeFiles/torch_hip.dir/utils/math/torch_hip_generated_transpose.cu.o
FAILED: caffe2/CMakeFiles/torch_hip.dir/utils/math/torch_hip_generated_transpose.cu.o
cd /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math && /usr/bin/cmake -E make_directory /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math/. && /usr/bin/cmake -D verbose:BOOL=OFF -D build_configuration:STRING=RELEASE -D generated_file:STRING=/home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math/./torch_hip_generated_transpose.cu.o -P /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math/torch_hip_generated_transpose.cu.o.cmake
In file included from /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/caffe2/utils/math/transpose.cu:7:
/home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/caffe2/core/common_gpu.h:5:10: fatal error: 'cuda.h' file not found
#include <cuda.h>
         ^~~~~~~~
1 error generated when compiling for host.
CMake Error at torch_hip_generated_transpose.cu.o.cmake:138 (message):
  Error generating
  /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math/./torch_hip_generated_transpose.cu.o


[3/231] Building HIPCC object caffe2/CMakeFiles/torch_hip.dir/core/torch_hip_generated_context_gpu.cu.o
FAILED: caffe2/CMakeFiles/torch_hip.dir/core/torch_hip_generated_context_gpu.cu.o
cd /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/core && /usr/bin/cmake -E make_directory /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/core/. && /usr/bin/cmake -D verbose:BOOL=OFF -D build_configuration:STRING=RELEASE -D generated_file:STRING=/home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/core/./torch_hip_generated_context_gpu.cu.o -P /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/core/torch_hip_generated_context_gpu.cu.o.cmake
In file included from /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/caffe2/core/context_gpu.cu:8:
In file included from /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/c10/cuda/CUDACachingAllocator.h:4:
/home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/c10/cuda/CUDAStream.h:6:10: fatal error: 'cuda_runtime_api.h' file not found
#include <cuda_runtime_api.h>
         ^~~~~~~~~~~~~~~~~~~~
1 error generated when compiling for host.
CMake Error at torch_hip_generated_context_gpu.cu.o.cmake:138 (message):
  Error generating
  /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/core/./torch_hip_generated_context_gpu.cu.o


[4/231] Building HIPCC object caffe2/CMakeFiles/torch_hip.dir/utils/math/torch_hip_generated_elementwise.cu.o
FAILED: caffe2/CMakeFiles/torch_hip.dir/utils/math/torch_hip_generated_elementwise.cu.o
cd /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math && /usr/bin/cmake -E make_directory /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math/. && /usr/bin/cmake -D verbose:BOOL=OFF -D build_configuration:STRING=RELEASE -D generated_file:STRING=/home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math/./torch_hip_generated_elementwise.cu.o -P /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math/torch_hip_generated_elementwise.cu.o.cmake
In file included from /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/caffe2/utils/math/elementwise.cu:10:
In file included from /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/caffe2/core/context_gpu.h:8:
/home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/caffe2/core/common_gpu.h:5:10: fatal error: 'cuda.h' file not found
#include <cuda.h>
         ^~~~~~~~
1 error generated when compiling for host.
CMake Error at torch_hip_generated_elementwise.cu.o.cmake:138 (message):
  Error generating
  /home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/build/caffe2/CMakeFiles/torch_hip.dir/utils/math/./torch_hip_generated_elementwise.cu.o


ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "setup.py", line 732, in <module>
    build_deps()
  File "setup.py", line 311, in build_deps
    build_caffe2(version=version,
  File "/home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/tools/build_pytorch_libs.py", line 62, in build_caffe2
    cmake.build(my_env)
  File "/home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/tools/setup_helpers/cmake.py", line 345, in build
    self.run(build_args, my_env)
  File "/home/acxz/vcs/git/github/rocm-arch/python-pytorch-rocm/src/pytorch-1.6.0-rocm/tools/setup_helpers/cmake.py", line 141, in run
    check_call(command, cwd=self.build_dir, env=env)
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--target', 'install', '--config', 'Release', '--', '-j', '4']' returned non-zero exit status 1.
1 acxz@archard ..on-pytorch-rocm/src/pytorch-1.6.
```

## To Reproduce

High level:
1. git clone
2. python tools/amd_build/build_amd.py
2. python setup.py build

Technically I am using the following PKGBUILD script [here](https://github.com/rocm-arch/python-pytorch-rocm/blob/master/PKGBUILD)

## Expected behavior

Should compile successfully

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.6.0
 - OS (e.g., Linux): Arch Linux
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): `python setup.py build`
 - Python version: 3.8.5
 - CUDA/cuDNN version: N/A
 - GPU models and configuration:
 - Any other relevant information: ROCm

## Additional context

Downstream issue: https://github.com/rocm-arch/python-pytorch-rocm/issues/5

cc @malfet @jeffdaily @sunway513 @t-vi

**Comment by t-vi** [comment]
So I don't run into this, even though I also have the include statement.
Two things that come to mind, even though I don't know if that's the root cause:
1. My ROCm (from AMDs debs) comes with an empty (well, one lf char) cuda.h somewhere (but I don't know if its in the search path):
```
$ ls -l /opt/rocm-3.5.0/hip/include/hip/hcc_detail/cuda/cuda.h
-rw-r--r-- 1 root root 1 May 22 01:54 /opt/rocm-3.5.0/hip/include/hip/hcc_detail/cuda/cuda.h
```
2. I actually do have `cuda.h` installed in `/usr/include/cuda.h`, though it would be bad if we relied on that.

**Comment by acxz** [comment]
Ah thanks for commenting on that. Yeah ROCm does provide a `cuda.h`. I have it in the same location as well on Arch Linux. hmm we need to analyze what files are being linked/included when building caffe2, do you know which files these would be? The problem is that `cuda.h` is just not in the include path.

Would you be willing to try to reproduce the error by removing your `/usr/include/cuda.h` file?

**Comment by acxz** [comment]
So the CI is using caffe2 from a docker with sccache? Not really sure how this is working, but it seems like pytorch might not even be building caffe2 but just getting it from somewhere else. I could try to replicate that part of the CI on my local script and see what happens.

https://github.com/pytorch/pytorch/blob/master/.jenkins/pytorch/build.sh#L134

**Comment by jeffdaily** [comment]
I'm trying to reproduce.  

Concerning sccache, ROCm builds do support it, and there are recent efforts to improve it.
- https://github.com/pytorch/pytorch/commit/79cdd84c81e8915b7ce0593a16f2ece305ec9641
- https://github.com/pytorch/pytorch/commit/2bc7dae2fc7ac09bb003973fdcfb76899a59b157

**Comment by jeffdaily** [comment]
I'm not able to reproduce the error.  I prepared a docker image based on archlinux and the rocm-arch packages (https://github.com/rocm-arch/rocm-arch).  I followed your steps rather than trying to use your PKGBUILD.  I did see an error for find(RCCL), and tried your patch (https://github.com/rocm-arch/python-pytorch-rocm/blob/master/find-rccl.patch), which allowed the build to proceed but failed to link (`/usr/bin/ld: cannot find -l__caffe2_nccl`).

**Comment by jeffdaily** [comment]
See my suggestion to your find rccl PR https://github.com/pytorch/pytorch/pull/42072.  It allowed me to finish my pytorch build.

**Comment by acxz** [comment]
Sweet, happy that you got pytorch building on Arch! Thx for your time in investigating this.
At least we got down to the rccl issue from this.

So I tried not using the PKGBUILD, and just the steps I listed. (with `USE_CUDA=OFF`) and I got further along but still ended up with this error: https://github.com/pytorch/pytorch/issues/42512

As for why the PKGBUILD is not working, I guess it is probably due to the fact that I have a bunch of CUDA flags in the PKGBUILD rn. I'll do some more local testing.

I think this issue can be closed considering I cannot reproduce it when building pytorch baremetal and the PKGBUILD issues are my own.
